### ISSUE TITLE
CRC error in BSplayer Pro

### ISSUE THREAD

**Issue by Vort** [bochs-emu/Bochs#372]
During installation of BSplayer Pro 1.22 Build 817 in Windows 95, such error appears:

![bochs_bsp](https://github.com/user-attachments/assets/ee8d5ba9-e9fb-4056-abbe-d9d3912746f5)

When installed into VirtualBox 6.1.36 r152435 from the same file, no such error appear.

I'm almost sure CRC is fine and program detects emulation incorrectness or inaccuracy somehow.

It may be that program just searches for Bochs-specific devices or something like that, in such cases it makes sense to close this issue without fixing.
Otherwise, if CPU emulation have mismatches, for example, it is better to fix them.

Test file: [BSPLAYER_PRO122.817.ZIP](https://github.com/user-attachments/files/17531661/BSPLAYER_PRO122.817.ZIP)

Version: a94ebf9

**Comment by stlintel** [comment]
For CPU behavior - try to test with different CPU model, someone older which always worked.
If it works without a error - it could be CP related.
Otherwise, most likely device or BIOS.

**Comment by Vort** [comment]
> For CPU behavior - try to test with different CPU model

I tested `pentium`, `core2_penryn_t9600` and `corei7_haswell_4770` - same effect.

> If it works without a error - it could be CP related.
Otherwise, most likely device or BIOS.

There also may be problem, which is shared by all CPU models.

Message mentions debugging, maybe debugging features of CPU are less tested.

**Comment by stlintel** [comment]
I doubt it is using any of them. It is probably generic message for most common issue in real life.
It indicates 'file content was modified'. Could it be that reading from disk is not working well ?

**Comment by Vort** [comment]
I noticed lags during HDD usage in the past (during Windows installation for example).
But I never saw data corruption.
You are right however - there may be lots of reasons for this message.

**Comment by Vort** [comment]
When player binary is unpacked with [tElock Unpacker.zip](https://github.com/user-attachments/files/17533219/tElock.Unpacker.zip), it starts normally, so it's the [tElock.0.98.zip](https://github.com/user-attachments/files/17533221/tElock.0.98.zip), which triggers problem.
So I assembled simple program, [pedemo.zip](https://github.com/user-attachments/files/17533222/pedemo.zip), and added tElock on top of it.
As a result, `pedemo.exe` shows nothing in Bochs, works good in VirtualBox _(and crashes on my host OS, Windows 7)_, while `pedemo_orig.exe` works as expected in all tests.
Please note, that your antiviruses may not like contents of these archives, this is frequent reaction to these kind of tools.

**Comment by Vort** [comment]
Surprisingly, tElock`ed binaries are compatible with WinAPIOverride.
Which means I can compare API calls between Bochs and VirtualBox (inside of Windows XP this time).
This is how it should work (VirtualBox):
![pedemo_virtualbox](https://github.com/user-attachments/assets/4ae4e1fe-25fb-4407-990a-e292952d42b2)
This is what happens (Bochs):
![pedemo_bochs](https://github.com/user-attachments/assets/573a8316-f824-4321-93a1-8b83c0c2cdee)

Something goes wrong before execution of `GetModuleHandleA("KERNEL32.DLL")`.

**Comment by Vort** [comment]
> I doubt it is using any of them.

I tried to use hardware breakpoints in Bochs with OllyDbg and they fail.
This is how CPU state should look at problematic place (screenshot was obtained with VirtualBox):
![olly_pedemo_405241](https://github.com/user-attachments/assets/62c715f6-e3ab-44a9-b72f-6e8423b42fd1)

Bochs runs pass this place despite breakpoint.
Theoretically, Bochs debugger can help, but I don't know how to use it for debugging of Windows programs.

**Comment by Vort** [comment]
@stlintel, here is the difference I found:

```asm
include 'win32ax.inc'
.code
start:
  mov  eax, 0x20A3
  ror  al, 7
  jo   bad
  invoke MessageBox, HWND_DESKTOP, "Good", "Result", MB_OK
  jmp exit
bad:
  invoke MessageBox, HWND_DESKTOP, "Bad", "Result", MB_OK
exit:
  invoke ExitProcess, 0
.end start
```

"Bad" result in Bochs, "Good" result on real hardware and in VirtualBox.

Test file: [rortest.zip](https://github.com/user-attachments/files/17538893/rortest.zip).

**Comment by stlintel** [comment]
Hi, this is grey area here - undefined flags.
The SDM clearly states that OF flag is undefined after ROR operation with count>1

`If the masked count is 1, then the OF flag is affected, otherwise  (masked count is greater than 1) the OF flag is undefined. `

Actual behavior is different between different processor generations.
Old processors (until Sandy Bridge - Intel Core 2xxx) use different rule from newer processors.
If your "real hardware" is newer than 2006: it should be newer rule.
But the Bochs rule was implemented before and it is still old rule.

OLD: 
	   ROTATE RIGHT RULE: OF = Provided the count is 1 or greater,
					XOR of 2 highest order bits of the destination.

NEW:
           SNB_ROR_OF:   if (count > 1), EFLAGS.OF = Src2[MSB] XOR Src2[0]

for your example:
 mov  eax, 0x20A3
 ror  al, 7

AL=0xA3 binary=1010.0011 after ROR: 010.00111
OLD rule:  result[7]^result[6] = 0^1=1 overflow
NEW rule: src[7]^src[0] = 1^1=0 no overflow


There are few more instructions which flags handling was changed starting from Sandy Bridge.
OF flag for the following instructions: SHL, SHR, SHRD, SHLD, ROR, RCR, ROL, RCL

Do we need to bother for these flags updates ?
Are you sure it is correct software ?
It it is new software - it would be breaking on old processors (like Pentium 4)
If it is old software - it clearly cannot rely on new behavior for the flags ...
Are you sure we need to

**Comment by Vort** [comment]
> Do we need to bother for these flags updates ?

It's the first difference I noticed during analysis of tElock.
I'm not sure if it causes error.
I will look more into it.

> BTW, emulation of HW breakpoints require to configure Bochs with:
--enable-x86-debugger

Actually, I was able to use data breakpoints successfully.
It's execution breakpoints which failed.

I also tried version with debugger, but probably missed something - will recheck it as well.

**Comment by Vort** [comment]
> Are you sure this was was the real issue ?

It was not.
Sorry, I thought CPU behaviour should match in this case.

Looks like real issue is actually somehow related to debugging features of CPU:
With `--enable-x86-debugger` build, CRC error just disappear.
_(I can fully debug this problem, but it is gone)_

**Comment by stlintel** [comment]
I guess the right solution will be to enable x86 debugger extensions by default. Once they were disabled because 
a) they were experimental and not debugged enough
b) they were assumed to affect simulation speed 
Both these concerns are not relevant/worked around long time ago. Today there is no problem to have it always enabled.

**Comment by Vort** [comment]
> a) they were experimental and not debugged enough

I though this feature was added relatively recently.
But after reading [x86 debug register](https://en.wikipedia.org/wiki/X86_debug_register) article, I realized that it is available since 386 processor.
I think it totally makes sense to have all 386 features enabled by default.
Possible bugs here can be catched in the same way as it happens with other subsystems.

> b) they were assumed to affect simulation speed

That was my main concern, but I made few tests with and without `--enable-x86-debugger` and I see almost no difference in emulation speed.
